pandas/tests/dtypes/cast/test_promote.py

```diff
@@ -0,0 +1,732 @@
+# -*- coding: utf-8 -*-
+
+"""
+These test the private routines in types/cast.py
+
+"""
+
+import numpy as np
+import pytest
+
+from pandas._libs.tslibs import iNaT
+from pandas.compat import string_types
+
+from pandas.core.dtypes.cast import maybe_promote
+from pandas.core.dtypes.common import is_datetime64tz_dtype
+
+
+def test_maybe_promote_nan(any_numpy_dtype):
+    dtype = np.dtype(any_numpy_dtype)
+
+    if dtype == bytes:
+        pytest.xfail('does not upcast correctly!')
+
+    # worst case (=default) is upcasting to object, unless one of
+    # the conditions below is satisfied
+    expected_dtype = np.dtype(object)
+    expected_na_value = np.nan
+
+    if (issubclass(dtype.type, (np.datetime64, np.timedelta64))
+            or is_datetime64tz_dtype(dtype)):
+        # timedelta would otherwise be subclass of np.integer below
+        expected_dtype = dtype
+        expected_na_value = iNaT
+    elif issubclass(dtype.type, np.integer):
+        expected_dtype = np.float64
+    elif issubclass(dtype.type, (np.floating, np.complexfloating)):
+        expected_dtype = dtype
+
+    result_dtype, result_na_value = maybe_promote(dtype, fill_value=np.nan)
+    assert result_dtype == expected_dtype
+
+    # iNaT == iNaT, but np.nan != np.nan
+    assert (expected_na_value == result_na_value
+            or (expected_na_value is np.nan and result_na_value is np.nan))
+
+
+def test_maybe_promote_none(any_numpy_dtype):
```

Comment:
this should be the same as the above case, we always promote to np.nan

Reply:
Yeah, this is a left-over from the currently very inconsistent behaviour of not sometimes returning the value of `fill_value`.